### DISCUSSION TITLE
How to generate migrations?

### DISCUSSION THREAD

**Discussion by oco-adam** [medusajs/medusa#3366] [Q&A]
Hi, I'm new to medusa and Typeorm - I'm trying to generate migrations for a custom medusa model by using Typeorm's generate functionality. 

In the [medusajs docs](https://docs.medusajs.com/advanced/backend/migrations/) it says 

_You can alternatively use Typeorm's generate command to generate a Migration file from existing database tables, which requires setting up a data source in Typeorm. Check out Typeorm's documentation to learn how to create a data source, then use the generate command._

So I looked in Typeorm's docs [here](https://typeorm.io/data-source#creating-a-new-datasource), but it is not clear to me where I should initialise this datasource and how then I can use it to generate migrations... 

I have tried adding a file in in my project's root dir called datasource.js with the following content:

```js
import { DataSource } from "typeorm";

export const AppDataSource = new DataSource({
  type: "postgres",
  host: process.env.DB_HOST,
  port: process.env.DB_PORT,
  username: process.env.DB_USERNAME,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_DATABASE,
  migrations: [
    "./node_modules/@medusajs/medusa/dist/migrations/*.js",
    "./dist/migrations/*.js",
  ],
  entities: [
    "./node_modules/@medusajs/medusa/dist/models/*.js",
    "./dist/models/*.js",
  ],
});

AppDataSource.initialize()
  .then(() => {
    console.log("Data Source has been initialized!");
  })
  .catch((err) => {
    console.error("Error during Data Source initialization", err);
  });
```

Then I try and generate a migration for my new custom model using the command:

```
npx typeorm migration:generate -n MyModel -d ./src/migrations/
```

but I'm getting the following error:

```
Error during migration generation:
TypeORMError: No connection options were found in any orm configuration files.
```

In the Typeorm online docs it says that -d should be the path to the datasource, but in the medusa docs and Typeorm command line tool help it says that -d indicates the directory where the migration should be created, so it seems like there is something strange going on...🧐

I've tried many different combinations (including with the option -d ./datasource.js) and none of these ways work for me. 

I'm wondering if anyone is currently using this TypeORM generate functionality with Medusa and if so then please can you give me an idea of where I'm going wrong?

**Comment by adrien2p** [comment]
As per the documentation the command should be
`typeorm  migration:generate -p -d datasource.js src/entities/post`

For reference https://orkhan.gitbook.io/typeorm/docs/migrations#

**Reply by adrien2p** [reply]
Here is an example of the datasource file in the root of my dummy project

```js
// datasource.js

const { DataSource } = require("typeorm");

const dataSourceOptions = {
  "type": "postgres",
  "url": "postgres://postgres@localhost/medusa",
  "extra": {},
  "entities": [
    "node_modules/@medusajs/medusa/dist/models/!(*.index.js)",
    "dist/entities/*.js"
  ],
  "migrations": [
    "node_modules/@medusajs/medusa/dist/migrations/*.js",
    "dist/migrations/*.js"
  ]
}

module.exports = {
  datasource: new DataSource(dataSourceOptions)
}
```

here is the entity

```ts
// src/models/post.ts

import { Column, Entity, PrimaryGeneratedColumn } from "typeorm"

@Entity()
export class Post {
    @PrimaryGeneratedColumn({ primaryKeyConstraintName: "PK_post_id" })
    id: number

    @Column()
    title: string

    @Column()
    text: string
}
```

then I run `yarn build`

finally as per the documentation I ve shared above `typeorm  migration:generate -p -d datasource.js src/entities/post`

and the output contains a bunch of stuff including 

```ts
        await queryRunner.query(`
            CREATE TABLE "post" (
                "id" SERIAL NOT NULL,
                "title" character varying NOT NULL,
                "text" character varying NOT NULL,
                CONSTRAINT "PK_post_id" PRIMARY KEY ("id")
            )
        `);
```

if In the datasoruce I remove the reference to meduas entities, the datasource become

```js
// datasource.js

const { DataSource } = require("typeorm");

const dataSourceOptions = {
  "type": "postgres",
  "url": "postgres://postgres@localhost/medusa",
  "extra": {},
  "entities": [
    // "node_modules/@medusajs/medusa/dist/models/!(*.index.js)",
    "dist/entities/*.js"
  ],
  "migrations": [
    "node_modules/@medusajs/medusa/dist/migrations/*.js",
    "dist/migrations/*.js"
  ]
}

module.exports = {
  datasource: new DataSource(dataSourceOptions)
}
```

and then re run `typeorm  migration:generate -p -d datasource.js src/entities/post`

my output become

```ts
import { MigrationInterface, QueryRunner } from "typeorm";

export class post1677885254028 implements MigrationInterface {
    name = 'post1677885254028'

    public async up(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`
            CREATE TABLE "post" (
                "id" SERIAL NOT NULL,
                "title" character varying NOT NULL,
                "text" character varying NOT NULL,
                CONSTRAINT "PK_post_id" PRIMARY KEY ("id")
            )
        `);
    }

    public async down(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`
            DROP TABLE "post"
        `);
    }

}
```

But if in your entity you create a relation with a medusa entity, I suppose that the first example is what you need. About the other stuff generated along side your table, I suppose that we have to do something on our side.

Hope that everything is clear, otherwise do not hesitate

**Comment by shahednasser** [comment]
Hi @oco-adam, thank you for posting this question. We updated the documentation with an example of how to generate migrations from entities. The answer @adrien2p left won't work in the current released version of Medusa, as Medusa currently uses v `0.2.45` of Typeorm which doesn't have data sources. We will be releasing an update soon of Medusa which will update Typeorm to the latest version, but for now you can try the steps to generate the migration as in the updated documentation:

https://docs.medusajs.com/advanced/backend/migrations/#step-1-create-migration-file
